universal-application-tool-0.0.1/app/services/export/ExporterService.java

```diff
@@ -150,12 +150,20 @@ public String exportCsv(CsvExporter csvExporter, ImmutableList<Application> appl
     try {
       OutputStream inMemoryBytes = new ByteArrayOutputStream();
       Writer writer = new OutputStreamWriter(inMemoryBytes, StandardCharsets.UTF_8);
+      HashMap<Long, ProgramDefinition> programDefinitions = new HashMap<>();
       for (Application application : applications) {
+        Long programId = application.getProgram().id;
+        if (!programDefinitions.containsKey(programId)) {
+          try {
+            programDefinitions.put(programId, programService.getProgramDefinition(programId));
+          } catch (ProgramNotFoundException e) {
+            throw new RuntimeException("Cannot find a program that has applications for it.", e);
+          }
+        }
```

Comment:
suggestion (non-blocking): load all programs in a single query rather than one query per program or leave a TODO and file an issue for doing that in the future.

Loading all of the programs needed for the export with a single DB query would be an improvement over loading them individually. That being said that query doesn't exist yet on `ProgramRepository` (though very similar queries do) and this is time sensitive.

Reply:
I'll do a TODO for that.

Reply:
TODOs for single lookup and Javacache